toys/slot-machine/tests/lib.rs

```diff
@@ -0,0 +1,86 @@
+use radix_engine::ledger::*;
+use radix_engine::model::*;
+use radix_engine::transaction::*;
+use scrypto::prelude::*;
+
+#[test]
+fn test_slot_machine() {
+    // Set up environment.
+    let mut ledger = InMemorySubstateStore::with_bootstrap();
+    let mut executor = TransactionExecutor::new(&mut ledger, true);
+    let key = executor.new_public_key();
+    let account = executor.new_account(key);
+    let package = executor.publish_package(include_code!("slot_machine")).unwrap();
+
+    // Test the `new` function.
+    let transaction1 = TransactionBuilder::new(&executor)
+        .call_function(
+            package,
+            "SlotMachine",
+            "new",
+            vec![
+                "100000".to_owned()
+            ],
+            None
+        )
+        .build(vec![key])
+        .unwrap();
+    let receipt1 = executor.run(transaction1).unwrap();
+    println!("{:?}\n", receipt1);
+    assert!(receipt1.result.is_ok());
+
+    // Test the `free_token` method.
+    let casino = receipt1.component(0).unwrap();
+    let vegas_token_ref = receipt1.resource_def(0).unwrap();
+    let amount = fungible_amount(10, vegas_token_ref);
+    let transaction2 = TransactionBuilder::new(&executor)
+        .call_method(casino, "free_token", vec![], Some(account))
+        .assert_worktop_contains(amount.amount().unwrap(), amount.resource_address())
+        .take_from_worktop(&amount, |builder, bid| {
+            builder
+                .add_instruction(Instruction::CallMethod {
+                    component_address: account,
+                    method: "deposit".to_owned(),
+                    args: vec![scrypto_encode(&bid)],
+                })
+                .0
+        })
+        .build(vec![key])
+        .unwrap();
+    let receipt2 = executor.run(transaction2).unwrap();
+    println!("{:?}\n", receipt2);
+    assert!(receipt2.result.is_ok());
+
+    // Test the `play` method.
+    let bet = Decimal::one();
+    let winnigs = fungible_amount(2, vegas_token_ref);
+    let transaction3 = TransactionBuilder::new(&executor)
```

Comment:
@cbisaillon So this play transaction randomly wins and leaves 2 tokens on the worktop which I take a deposit to the players account. I don't think it is really maintainable because with the next scrypto update the play might return a different result.

Is there currently a way to tell "Take whatever amount is on the worktop and place it the players account."?

Reply:
You can do `call_method_with_all_resources`:

```
let transaction2 = TransactionBuilder::new(&executor)
        .call_method(component, "free_token", vec![], Some(account))
        .call_method_with_all_resources(account, "deposit_batch")
        .build(vec![key])
        .unwrap();
```

Reply:
Thanks!